src/lib.rs

```diff
@@ -1,7 +1,1242 @@
+        ChildIndex::from_index(0)
+    }
+
+    fn to_index(&self) -> u32 {
+        match self {
+            &ChildIndex::Hardened(i) => i + (1 << 31),
+            &ChildIndex::NonHardened(i) => i,
+        }
+    }
+}
+
+/// A chain code
+#[derive(Clone, Copy, Debug, PartialEq)]
+struct ChainCode([u8; 32]);
+
+#[derive(Clone, Copy, Debug, PartialEq)]
+pub struct DiversifierIndex(pub [u8; 11]);
+
+impl DiversifierIndex {
+    fn new() -> Self {
+        DiversifierIndex([0; 11])
+    }
+
+    pub fn increment(&mut self) -> Result<(), ()> {
+        let mut k = 0;
+        loop {
+            self.0[k] = self.0[k].wrapping_add(1);
+            if self.0[k] != 0 {
+                // No overflow
+                return Ok(());
+            }
+            // Overflow
+            k += 1;
+            if k == 11 {
+                return Err(());
+            }
+        }
+    }
+}
+
+/// A key used to derive diversifiers for a particular child key
+#[derive(Clone, Copy, Debug, PartialEq)]
+struct DiversifierKey([u8; 32]);
+
+impl DiversifierKey {
+    fn master(sk_m: &[u8]) -> Self {
+        let mut dk_m = [0u8; 32];
+        dk_m.copy_from_slice(&prf_expand(sk_m, &[0x10]).as_bytes()[..32]);
+        DiversifierKey(dk_m)
+    }
+
+    fn derive_child(&self, i_l: &[u8]) -> Self {
+        let mut dk = [0u8; 32];
+        dk.copy_from_slice(&prf_expand_vec(i_l, &[&[0x16], &self.0]).as_bytes()[..32]);
+        DiversifierKey(dk)
+    }
+
+    /// Returns the first index starting from j that generates a valid
+    /// diversifier, along with the corresponding diversifier. Returns
+    /// an error if the diversifier space is exhausted.
+    fn diversifier(&self, mut j: DiversifierIndex) -> Result<(DiversifierIndex, Diversifier), ()> {
+        let ff = FF1::<Aes256>::new(&self.0, 2).unwrap();
+        loop {
+            // Generate d_j
+            let enc = ff.encrypt(&[], &BinaryNumeralString::from_bytes_le(&j.0[..]))
+                .unwrap();
+            let mut d_j = [0; 11];
+            d_j.copy_from_slice(&enc.to_bytes_le());
+            let d_j = Diversifier(d_j);
+
+            // Return (j, d_j) if valid, else increment j and try again
+            match d_j.g_d::<Bls12>(&JUBJUB) {
+                Some(_) => return Ok((j, d_j)),
+                None => if j.increment().is_err() {
+                    return Err(());
+                },
+            }
+        }
+    }
+}
+
+/// A Sapling extended spending key
+#[derive(Clone)]
+pub struct ExtendedSpendingKey {
+    depth: u8,
+    parent_fvk_tag: FVKTag,
+    child_index: ChildIndex,
+    chain_code: ChainCode,
+    pub expsk: ExpandedSpendingKey<Bls12>,
+    dk: DiversifierKey,
+}
+
+// A Sapling extended full viewing key
+pub struct ExtendedFullViewingKey {
+    depth: u8,
+    parent_fvk_tag: FVKTag,
+    child_index: ChildIndex,
+    chain_code: ChainCode,
+    pub fvk: FullViewingKey<Bls12>,
+    dk: DiversifierKey,
+}
+
+impl std::cmp::PartialEq for ExtendedSpendingKey {
```

Comment:
Why can we not #[derive(PartialEq)] for this/the structs it contains?

Reply:
`#[derive(PartialEq)]` only works if all the inner fields of the struct implement `PartialEq`, which is not the case here.

Reply:
It might be nice to derive PartialEq for the types of the inner fields also. Would it make sense to do that? Not that we've done anything wrong, but if we are just doing pairwise comparison of the fields it might be nice it the compiler were in charge of that rather than us.

Reply:
That requires changes in sapling-crypto, and also spirals off into pairing IIRC due to the way derives handle generic parameters. For now, this is much simpler.